### PR TITLE
Add x86_64-unknown-linux-none target

### PR DESCRIPTION
**Pull request by rosymati**
Adds a freestanding linux binary with no libc dependency. This is useful for writing programs written only in rust. It is also essential for writing low level stuff like libc or a dynamic linker.

Tier 3 policy:

>A tier 3 target must have a designated developer or developers (the "target maintainers") on record to be CCed when issues arise regarding the target. (The mechanism to track and CC such developers may evolve over time.)

I will be the designed maintainer for this target

>Targets must use naming consistent with any existing targets; for instance, a target for the same CPU or OS as an existing Rust target should use the same name for that CPU or OS. Targets should normally use the same names and naming conventions as used elsewhere in the broader ecosystem beyond Rust (such as in other toolchains), unless they have a very good reason to diverge. Changing the name of a target can be highly disruptive, especially once the target reaches a higher tier, so getting the name right is important even for a tier 3 target.

The target triple is consistent with other targets

>Target names should not introduce undue confusion or ambiguity unless absolutely necessary to maintain ecosystem compatibility. For example, if the name of the target makes people extremely likely to form incorrect beliefs about what it targets, the name should be changed or augmented to disambiguate it.
If possible, use only letters, numbers, dashes and underscores for the name. Periods (.) are known to cause issues in Cargo.

There is no confusion with other targets since it explicitly adds "none" at the end instead of omitting the environment

>Tier 3 targets may have unusual requirements to build or use, but must not create legal issues or impose onerous legal terms for the Rust project or for Rust developers or users.

The target does not introduce any unusual requirement

>The target must not introduce license incompatibilities.

There are no license incompatibilities

> Anything added to the Rust repository must be under the standard Rust license (MIT OR Apache-2.0).

Everything added is under that license

>The target must not cause the Rust tools or libraries built for any other host (even when supporting cross-compilation to the target) to depend on any new dependency less permissive than the Rust licensing policy. This applies whether the dependency is a Rust crate that would require adding new license exceptions (as specified by the tidy tool in the rust-lang/rust repository), or whether the dependency is a native library or binary. In other words, the introduction of the target must not cause a user installing or running a version of Rust or the Rust tools to be subject to any new license requirements.

There are no new dependencies

>Compiling, linking, and emitting functional binaries, libraries, or other code for the target (whether hosted on the target itself or cross-compiling from another target) must not depend on proprietary (non-FOSS) libraries. Host tools built for the target itself may depend on the ordinary runtime libraries supplied by the platform and commonly used by other applications built for the target, but those libraries must not be required for code generation for the target; cross-compilation to the target must not require such libraries at all. For instance, rustc built for the target may depend on a common proprietary C runtime library or console output library, but must not depend on a proprietary code generation library or code optimization library. Rust's license permits such combinations, but the Rust project has no interest in maintaining such combinations within the scope of Rust itself, even at tier 3.

There is no proprietary dependencies

>"onerous" here is an intentionally subjective term. At a minimum, "onerous" legal/licensing terms include but are not limited to: non-disclosure requirements, non-compete requirements, contributor license agreements (CLAs) or equivalent, "non-commercial"/"research-only"/etc terms, requirements conditional on the employer or employment of any particular Rust developers, revocable terms, any requirements that create liability for the Rust project or its developers or users, or any requirements that adversely affect the livelihood or prospects of the Rust project or its developers or users.

No such terms exist for this target

>Neither this policy nor any decisions made regarding targets shall create any binding agreement or estoppel by any party. If any member of an approving Rust team serves as one of the maintainers of a target, or has any legal or employment requirement (explicit or implicit) that might affect their decisions regarding a target, they must recuse themselves from any approval decisions regarding the target's tier status, though they may otherwise participate in discussions.

>This requirement does not prevent part or all of this policy from being cited in an explicit contract or work agreement (e.g. to implement or maintain support for a target). This requirement exists to ensure that a developer or team responsible for reviewing and approving a target does not face any legal threats or obligations that would prevent them from freely exercising their judgment in such approval, even if such judgment involves subjective matters or goes beyond the letter of these requirements.

Understood    

>Tier 3 targets should attempt to implement as much of the standard libraries as possible and appropriate (core for most targets, alloc for targets that can support dynamic memory allocation, std for targets with an operating system or equivalent layer of system-provided functionality), but may leave some code unimplemented (either unavailable or stubbed out as appropriate), whether because the target makes it impossible to implement or challenging to implement. The authors of pull requests are not obligated to avoid calling any portions of the standard library on the basis of a tier 3 target not implementing those portions.

The target already implements core. It might be possible in the future to add support for alloc and std by leveraging crates such as [origin](https://github.com/sunfishcode/origin/) and [rustix](https://github.com/bytecodealliance/rustix)

> The target must provide documentation for the Rust community explaining how to build for the target, using cross-compilation if possible. If the target supports running binaries, or running tests (even if they do not pass), the documentation must explain how to run such binaries or tests for the target, using emulation if possible or dedicated hardware if necessary.

I believe the proper docs are added

>Tier 3 targets must not impose burden on the authors of pull requests, or other developers in the community, to maintain the target. In particular, do not post comments (automated or manual) on a PR that derail or suggest a block on the PR based on a tier 3 target. Do not send automated messages or notifications (via any medium, including via @) to a PR author or others involved with a PR regarding a tier 3 target, unless they have opted into such messages.
 
> Backlinks such as those generated by the issue/PR tracker when linking to an issue or PR are not considered a violation of this policy, within reason. However, such messages (even on a separate repository) must not generate notifications to anyone involved with a PR who has not requested such notifications.

Understood

> Patches adding or updating tier 3 targets must not break any existing tier 2 or tier 1 target, and must not knowingly break another tier 3 target without approval of either the compiler team or the maintainers of the other tier 3 target.

> In particular, this may come up when working on closely related targets, such as variations of the same architecture with different features. Avoid introducing unconditional uses of features that another variation of the target may not have; use conditional compilation or runtime detection, as appropriate, to let each target run code supported by that target.

No other targets are effected

>Tier 3 targets must be able to produce assembly using at least one of rustc's supported backends from any host target.

The same backends used by other linux targets work without issues

### CODE REVIEW

**Comment by morr0ne** [comment]
r? compiler-team

**Comment by GuillaumeGomez** [comment]
r? nielstrieb

**Comment by GuillaumeGomez** [comment]
r? nilstrieb

**Comment by GuillaumeGomez** [comment]
Arf didn't see that the reviewer was already changed... Putting back estebank.

r? estebank

**Comment by morr0ne** [comment]
Removed the merge commit, sorry about that

**Comment by workingjubilee** [comment]
@morr0ne could you more fully explain the intended distinction between this target and the `x86_64-unknown-none` and `x86_86-unknown-linux-gnu` (with `#![no_std]`) targets?

**Comment by workingjubilee** [comment]
> The target already implements core. It might be possible in the future to add support for alloc and std by leveraging crates such as [origin](https://github.com/sunfishcode/origin/) and [rustix](https://github.com/bytecodealliance/rustix)

More particularly, I believe it should be clear from day 0 what the stdlib impls are supposed to be.

**Comment by morr0ne** [comment]
> @morr0ne could you more fully explain the intended distinction between this target and the `x86_64-unknown-none` and `x86_86-unknown-linux-gnu` (with `#![no_std]`) targets?

The existing *-gnu targets make the assumption that the target system is running a gnu environment, which in pratice means assuming glibc and the corresponding linker are present. This new target is intended for targets that make zero assumption about the environment beside the underlining kernel (linux) conventions. 
The intended use cases for this targets are mainly 2 at the moment:
- Writing fully rust binaries that do not depend on libc
- Writing low level code such a libc implementation itself

I have been using this target in my own projects for months at this point and the benefit is clear. Without this target there are workarounds to achieve the same goals but they piggyback off the existing gnu targets. This way we have a target that clearly states that we are running without a libc

> More particularly, I believe it should be clear from day 0 what the stdlib impls are supposed to be. It is not required that those implementations should be done, and it is fully possible for those to vary in details on the journey to implementation, but I believe it makes it much harder to follow platform support processes if "I intend to" is hedged into "might be possible maybe someday or not".

Maybe I could have worded it better. Having std and alloc for this target is 100% possible, I have zero doubts about that. The "might" stems from me not knowing if that is within the scope of rust since that would require shipping an allocator with std. I can't confidently say "this target will support std and alloc" because I don't know if the rust projects wants such support for an enviroment-less target. What I can say is that if there's consensus and interest, it is possible to implement std and alloc

**Comment by morr0ne** [comment]
To further expand on my previous comment. The *-gnu targets assume the compiled binary will run on a GNU/Linux system. This new target is meant to target only Linux itself.

**Comment by workingjubilee** [comment]
> Maybe I could have worded it better. Having std and alloc for this target is 100% possible, I have zero doubts about that. The "might" stems from me not knowing if that is within the scope of rust since that would require shipping an allocator with std. I can't confidently say "this target will support std and alloc" because I don't know if the rust projects wants such support for an environment-less target. What I can say is that if there's consensus and interest, it is possible to implement std and alloc

Aha, okay! So you would be be happy to have this target with such implementations but also would be reasonably content without such implementations. Thank you, that clarifies the intent.

**Comment by workingjubilee** [comment]
I believe n order for this to land, you will want to edit ./tests/assembly/targets/targets-elf.rs so that it has a new entry for this platform that resembles these:

https://github.com/rust-lang/rust/blob/78a77512702ab786f6f9345872d36d852454612c/tests/assembly/targets/targets-elf.rs#L546-L557

**Comment by morr0ne** [comment]
> Aha, okay! So you would be be happy to have this target with such implementations but also would be reasonably content without such implementations. Thank you, that clarifies the intent.

Yes exactly

> I believe n order for this to land, beyond any policy questions that may need to be addressed, you will want to edit ./tests/assembly/targets/targets-elf.rs so that it has a new entry for this platform that resembles these:

I did see the ci is failing because of missing test. Perhaps I missed it in somewhere but I didn't see it documented in the guide I was following.

**Comment by workingjubilee** [comment]
Can you mention which guide you were following? We have a "do you emit the binary we were expecting" test for targets, even tier 3 ones, ever since a target once started to silently fail to produce the correct kind of binary overnight.

**Comment by morr0ne** [comment]
> Can you mention which guide you were following?

I used this documentation when I originally wrote the target: https://rustc-dev-guide.rust-lang.org/building/new-target.html
And this one to open the pr: https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html#adding-a-new-target

**Comment by workingjubilee** [comment]
It's mentioned in the latter:
> Tier 3 targets must be able to produce assembly using at least one of rustc's supported backends from any host target.

But the directive to update the test in question can be added to the rustc-dev-guide.

File: src/doc/rustc/src/platform-support/x86_64-unknown-linux-none.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# `x86_64-unknown-linux-none`
+
+**Tier: 3**
+
+Freestanding x86-64 linux binary with no depedency on libc.
```
Comment (bjorn3) [inline src/doc/rustc/src/platform-support/x86_64-unknown-linux-none.md:5]: *dependency

**Comment by morr0ne** [comment]
At the moment I can't really say I have strong requirements for a version. To actually be useful without a libc it definitely needs to rely on more modern versions of the kernel, at least compared to the existing *-gnu targets. It is hard to pin-point a specific version however and I am happy to leave it somewhat unspecified, at least until there is a clear path forward to implement std and alloc.

Generally speaking this target can be assumed to be somewhat bleeding edge. At the moment I work with the latest stable release of the kernel so it would be hard to notice if something is broken on older kernels

**Comment by estebank** [comment]
@Nilstrieb do you want to snipe this review? I won't be around for a few days and this seems out of my wheelhouse. @workingjubilee thank you for taking on digging deeper on this!

**Comment by briansmith** [comment]
> At the moment I can't really say I have strong requirements for a version. To actually be useful without a libc it definitely needs to rely on more modern versions of the kernel, at least compared to the existing *-gnu targets. It is hard to pin-point a specific version however and I am happy to leave it somewhat unspecified, at least until there is a clear path forward to implement std and alloc.

It would be good to document that an unknown kernel version, which is known to be greater than 3.19, because this affects whether `getrandom` can be used.

On Linux, it is normally required to route all direct syscall calls through libc's `syscall(2)`. Without a libc for this target, does this mean that we're supposed to issue syscalls directly, without any coordination from the runtime system? And further, are we guaranteed that the process will never be linked to any libc that requires syscalls to be routed through it? I think this is important to specify because there's no point in having target_os=linux without the ability to actually use Linux features.

In particular, how would we ask the OS to allocate memory for us to implement a heap?

What is the minimum CPU feature set for this target? Since we cannot use (IIUC) libstd's `std::arch::is_x86_feature_detected`, are we allowed to use the results of CPUID + xgetbv to dynamically determine which CPU features are available?

File: src/doc/rustc/src/platform-support/x86_64-unknown-linux-none.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+## Target maintainers
+
+- [morr0ne](https://github.com/morr0ne/)
+
+## Requirements
+
+This target is cross compiled and can be built from any host.
+
+This target has no support for host tools, std and alloc.
```
Comment (tgross35) [inline src/doc/rustc/src/platform-support/x86_64-unknown-linux-none.md:15]:
```suggestion
This target has no support for host tools, std, or alloc.
```

**Comment by tgross35** [comment]
Some background discussion at https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/Direct.20syscalls.20on.20Linux. Having this target will be awesome.

**Comment by morr0ne** [comment]
> It would be good to document that an unknown kernel version, which is known to be greater than 3.19, because this affects whether `getrandom` can be used.

It could be documented as such. But unless we have std we actually have no guarantee of such version. We might find out we actually can use a lower version, although unlikely.

> On Linux, it is normally required to route all direct syscall calls through libc's `syscall(2)`. Without a libc for this target, does this mean that we're supposed to issue syscalls directly, without any coordination from the runtime system?

Linux actually is the only mainstream os/kernel that allows us to make syscall directly. The userspace abi is stable and there is no standard libc. In a sense, there is no "runtime system" to coordinate with. There are many ways to make such syscalls but I believe the best solution is rustix, which I linked above.

> And further, are we guaranteed that the process will never be linked to any libc that requires syscalls to be routed through it?

The point of this target is exactly to guarantee no libc is linked. Linux doesn't forces you to link to a libc implementation and even if that were the case, there is no issue calling direct syscalls even if linking to libc. There is no requirement to make syscalls from libc on linux, regardless of the target environment.

> I think this is important to specify because there's no point in having target_os=linux without the ability to actually use Linux features.

The point of the target is to guarantee that we are running on top of the linux kernel. How syscalls are then made is up to the user in a no_std environments or up to std, however it decides to implement such calls.

> In particular, how would we ask the OS to allocate memory for us to implement a heap?

We can ask the os via syscalls the same way the libc implementation does. Unlike other OSes, libc is not special on linux. There is nothing libc can do that we cannot do ourselves by calling direct syscalls or reading special virtual filesystems like /proc

> What is the minimum CPU feature set for this target? Since we cannot use (IIUC) libstd's `std::arch::is_x86_feature_detected`, are we allowed to use the results of CPUID + xgetbv to dynamically determine which CPU features are available?

I do not know how std::arch::is_x86_feature_detected is implemented but I assume it is calling some libc function that can be replaced by our own implementation

**Comment by Nilstrieb** [comment]
> There is no requirement to make syscalls from libc on linux, regardless of the target environment.

This is not entirely true, libcs generally do not appreciate you calling some syscalls directly without talking to them first. This is of course a non-issue when the libc is not in the address space at all, but when it is linked in some other way, that can lead to problems (one example being the forking syscalls). I assume that's what the person meant.

Adding this target as a tier 3 no_std target makes sense. Any library implementation for this target should go to the libs team first to decide whether this is something that is desired in the first place.

Given that the target won't really rely on Linux kernel features for now, documenting the minimum version doesn't make that much sense imo - it's as low as anyone wants (though you can write something there if you feel like it). It would of course also need to be specified when adding a library implementation or relying on features.

**Comment by morr0ne** [comment]
> This is not entirely true, libcs generally do not appreciate you calling some syscalls directly without talking to them first. This is of course a non-issue when the libc is not in the address space at all, but when it is linked in some other way, that can lead to problems (one example being the forking syscalls). I assume that's what the person meant.

Maybe I missunderstood what they meant. Regardless this is a non issue since libc will not be linked in any way for this target.

> Adding this target as a tier 3 no_std target makes sense. Any library implementation for this target should go to the libs team first to decide whether this is something that is desired in the first place.

That was exactly my thought process. Adding this as a tier 3 makes this easy to merge without issues and any library implementation can be properly discussed at a later point.

> Given that the target won't really rely on Linux kernel features for now, documenting the minimum version doesn't make that much sense imo - it's as low as anyone wants (though you can write something there if you feel like it). It would of course also need to be specified when adding a library implementation or relying on features.

I fully agree with this and is exactly why I did not add a version in the documentation.

**Comment by Nilstrieb** [comment]
In that case, this is good to merge.

I will approve this PR, adding a new tier 3 no_std target for bare Linux without libc. It does not commit to a rustix (or other)-based implementation of the standard library for bare Linux in the future (this should be discussed with the libs team first).
As a tier 3 target, no builds are available and `-Zbuild-std` should be used. As usual, targets do not come with stability guarantees and may be removed in the future, especially tier 3 targets.

@bors r+

**Comment by programmerjake** [comment]
for things like `fork` and `brk` where libc expects you to call its functions instead of making syscalls, maybe you could add some dummy `fork`/`brk`/etc. functions to generate a linker error if you inadvertantly linked to libc.

**Comment by Nilstrieb** [comment]
no rustbot, that's noise

**Comment by Kobzol** [comment]
Noise, this shouldn't have changed perf.

@rustbot label: +perf-regression-triaged

**Comment by workingjubilee** [comment]
Yeah, for the purpose of the version, I literally asked so that if some time months down the line someone has to archaeology up the original PR for this target, that there is indeed an intention captured there regarding what version was intended. It's of course fine to not write it down.

> I do not know how `std::arch::is_x86_feature_detected` is implemented but I assume it is calling some libc function that can be replaced by our own implementation

@morr0ne For x86-64, it actually is entirely assembly.

**Comment by morr0ne** [comment]
> @morr0ne For x86-64, it actually is entirely assembly.

That is actually great to hear. It means we don't need to worry about the implementation

**Comment by williamdes** [comment]
If I am reading this good news correctly it means that I can build pure Rust programs to this target and have Docker images use `FROM scratch` and copy the binary into it. Until now most images are `FROM alpine` or `FROM debian`.
But sometimes it is better to have nothing to depend onto. 
Let me know if I understand this correctly

**Comment by morr0ne** [comment]
> If I am reading this good news correctly it means that I can build pure Rust programs to this target and have Docker images use `FROM scratch` and copy the binary into it. Until now most images are `FROM alpine` or `FROM debian`. But sometimes it is better to have nothing to depend onto. Let me know if I understand this correctly

Yes and no. The ultimate objective is to achieve fully independent rust programs, but this is an initial step since there is no std  or alloc implementation available. This target does support completely independent binaries, but as of now, it requires assistance from crates like `origin` or necessitates writing the startup code manually. Additionally, development must be carried out in a `no_std` environment.

Nevertheless, it is feasible to use this target in the manner you described.

**Comment by tgross35** [comment]
@morr0ne could you open a tracking issue for the `x86_64-unknown-linux-none` target if it doesn't exist already? I think a lot of people are interested but don't know what works and what needs work.

**Pull request merged**